pkg/distributor/distributor_ring.go

```diff
@@ -0,0 +1,87 @@
+package distributor
+
+import (
+	"flag"
+	"os"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/ring"
+	"github.com/cortexproject/cortex/pkg/ring/kv"
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/cortexproject/cortex/pkg/util/flagext"
+	"github.com/go-kit/kit/log/level"
+)
+
+// RingConfig masks the ring lifecycler config which contains
+// many options not really required by the distributors ring. This config
+// is used to strip down the config to the minimum, and avoid confusion
+// to the user.
+type RingConfig struct {
+	KVStore          kv.Config     `yaml:"kvstore,omitempty"`
+	HeartbeatPeriod  time.Duration `yaml:"heartbeat_period,omitempty"`
+	HeartbeatTimeout time.Duration `yaml:"heartbeat_timeout,omitempty"`
+
+	// Instance details
+	InstanceID             string   `yaml:"instance_id"`
+	InstanceInterfaceNames []string `yaml:"instance_interface_names"`
+	InstancePort           int      `yaml:"instance_port"`
+	InstanceAddr           string   `yaml:"instance_addr"`
+
+	// Injected internally
+	ListenPort int `yaml:"-"`
+}
+
+// RegisterFlags adds the flags required to config this to the given FlagSet
+func (cfg *RingConfig) RegisterFlags(f *flag.FlagSet) {
```

Comment:
Something similar for this would be nice for the ruler. The main difference being the number of tokens.

Reply:
Right. I would suggest to keep ruler's refactoring as a separate PR, given it's not related to this work. _I will add it to my backlog._